### ISSUE TITLE
Unable to use Bitbucket SSO

### ISSUE THREAD

**Issue by akfroman** [GitCredentialManager/git-credential-manager#329]
**Which version of GCM Core are you using?**
The version command does not work for me. I am using the version bundled with the Git 2.31.1 Windows installer.

**Which Git host provider are you trying to connect to?**

* [ ] Azure DevOps
* [ ] Azure DevOps Server (TFS/on-prem)
* [ ] GitHub
* [ ] GitHub Enterprise
* [X] Bitbucket
* [ ] Other - please describe

**Can you access the remote repository directly in the browser using the remote URL?**

From a terminal, run `git remote -v` to see your remote URL.

* [X] Yes
* [ ] No, I get a permission error
* [ ] No, for a different reason - please describe

---

**Expected behavior**

Bitbucket will open in my browser and allow me to authenticate using SSO. This is the behavior on my fresh Windows install machine.

**Actual behavior**

On my machine that previously had GCM for Windows installed, I uninstalled Git and GCM Windows, then reinstalled Git and checked the install GCM Core box. When I go to clone a repo from bitbucket, the small Bitbucket window pops up asking for a username/password (Like it did with GCM Windows) rather than opening my browser and allowing me to use SSO.

I believe there is some remanence of GCM Windows that didn't get uninstalled since this doesn't happen on my machine that never had GCM Windows on it. I've tried deleting all my git related credentials from Windows credential manager, but that didn't work.

I looked at the migration.md file. I ran `git config --list` and these are the only lines that begin with `credential.`
`credential.helper=manager-core`
`credential.helper=`
`credential.helper=C:/Program\ Files/Git/mingw64/libexec/git-core/git-credential-manager-core`

**Logs**

Set the environment variables `GCM_TRACE=1` and `GIT_TRACE=1` and re-run your Git command. Review and redact any private information and attach the log.

[log.txt](https://github.com/microsoft/Git-Credential-Manager-Core/files/6359184/log.txt)

**Comment by ibauersachs** [comment]
Are you talking about Bitbucket Cloud or Bitbucket Server? If the latter, then this is a duplicate of #326.

**Comment by akfroman** [comment]
This is for Bitbucket Cloud

**Comment by mminns** [comment]
Hi

When the username/password dialog opens do you just close it or do you enter your username/password?

**Comment by akfroman** [comment]
At first I closed it, but then I had to create an App password and I used that the next time

**Comment by mminns** [comment]
It is possible it is working correctly, it's been a while since I ran it.

See https://github.com/microsoft/Git-Credential-Manager-Core/blob/main/src/shared/Atlassian.Bitbucket/BitbucketHostProvider.cs#L123

Essentially, assuming no stored credentials, the flow is:

- Open a dialog and ask for username/password
- Use the u/p to try and access the user profile page in Bitbucket
- If the response is 200 the credentials are good and can be stored and used for git operations
- If the response is 403 then it means the user has 2FA turned on and must go through OAuth and GMC will open a dialog to run the OAuth flow.

**Comment by akfroman** [comment]
If I try to login with invalid credentials to invoke 403, all I get is `fatal: Invalid credentials` in the terminal and then it prompts me for my password in the terminal. If I enter an invalid password in the terminal I get 
```
remote: Invalid username or password
fatal: Authentication failed for <my repo>
```

**Comment by mminns** [comment]
Don't use invalid credentials.

Use your real Bitbucket username/password.

**Comment by akfroman** [comment]
Because of SSO, I don't have a password. I can create an App specific password, but I would like to avoid that.

**Comment by mminns** [comment]
Ah 💡 .

I see.

**Comment by mminns** [comment]
Its interesting that you imply you get different behaviour with GCMC compared to GCMW, do you think that?

The logic I described is the same for GCMC and GCMW.


I think the only way to get around this would be to add a configuration option like this https://github.com/microsoft/Git-Credential-Manager-Core/blob/d9d62c7ce9100c9b1866a5ca3ed390151ff6e7be/docs/configuration.md#credentialgitHubAuthModes

And use it to force OAuth only for Bitbucket

git config --global credential.bitbucketAuthModes "oauth"

**Comment by mminns** [comment]
@mjcheetham could you assign this one to me please?

**Comment by mminns** [comment]
I need to do some manual testing but I've got a prototype here https://github.com/microsoft/Git-Credential-Manager-Core/compare/main...mminns:ISSUE-329-force-bitbucket-oauth?expand=1

**Comment by mminns** [comment]
@akfroman please take a look at https://github.com/microsoft/Git-Credential-Manager-Core/pull/401.

I need to do a further test against Bitbucket Server/DC instances but it appears to be working correctly for me against https://bitbucket.org.

e.g. using

`git config --global credential.bitbucketAuthModes "oauth"` 

It skips the username/password Basic Auth step and takes you directly to the OAuth flow which I believe should be compatible with your SSO.

Feedback always welcome.

**Comment by mjcheetham** [comment]
Closing this issue now since the ability to force OAuth over user/pass for Bitbucket hosts has been released back in [v2.0.567](https://github.com/GitCredentialManager/git-credential-manager/releases/tag/v2.0.567).

Updating to the [latest Git for Windows](https://git-scm.com/downloads) and setting the `credential.bitbucketAuthModes` option should make GCM go straight to OAuth:

```batch
git config --global credential.bitbucketAuthModes "oauth"
```
